posts/2020-10-02-the-tale-about-hibernate-and-temporal-tables.adoc

```diff
@@ -0,0 +1,149 @@
+)
+WITH (SYSTEM_VERSIONING = ON (HISTORY_TABLE = dbo.DepartmentHistory));
+----
+
+This example comes from official documentation about temporal tables: https://docs.microsoft.com/en-us/sql/relational-databases/tables/creating-a-system-versioned-temporal-table?view=sql-server-ver15
+
+Then I add some records to our regular table:
+[source,sql]
+----
+insert into department (deptid, deptname) values (1, 'test1');
+insert into department (deptid, deptname) values (2, 'test2');
+----
+And update some records so history table will also get filled:
+[source,sql]
+----
+update department set deptname = 'updated' where deptid = 1;
+update department set deptname = 'updated_again' where deptid = 1;
+update department set deptname = 'updated2' where deptid = 2;
+----
+if we query the history table:
+[source,sql]
+----
+select * from departmenthistory;
+----
+we will see information like this:
+
+image::../media/2020-10-02-the-tale-about-hibernate-and-temporal-tables/db_result_1.png[]
+
+The department with DeptId = 1 was updated twice, so both old values are here and department with DeptId = 2 was updated once, so one old value of this record is also in the history table now.
+
+And if we want to query all history records for department with DeptId = 1 we will do this:
+[source,sql]
+----
+select * from departmenthistory where deptid = 1;
+----
+And will get this result:
+
+image::../media/2020-10-02-the-tale-about-hibernate-and-temporal-tables/db_result_2.png[]
+
+Now let's use Hibernate, just a simple small piece of code to run a native query:
+[source,java]
+----
+Query q = entityManager.createNativeQuery("select DeptId, DeptName from DepartmentHistory where DeptId = 1");
+----
+The result will return the following (which completely makes sense):
+[source,java]
+----
+1 test1 
+1 updated
+----
+Then we can create an Entity to easily operate with the data:
+[source,java]
+----
+@javax.persistence.Entity
+@Table(name = "DepartmentHistory")
+public class DepartmentHistory {
+
+    @Id
+    @Column(name = "DeptId")
+    private int DeptId;
+
+    @Column(name="DeptName")
+    private String DeptName;
+
+// other fields and getters and setters…
+
+}
+----
+and map our result properly to our entity:
+[source,java]
+----
+Query q = entityManager
+    .createNativeQuery("select * from DepartmentHistory where DeptId = 1", 
+    DepartmentHistory.class);
+----
+We expect to get the same result, but this is what we get instead:
+[source,java]
+----
+1 test1
+1 test1
+----
+So we get two rows, but they contain the same values. If there were more rows, they would all return the same data. This happens because of hibernate First level caching. Because all rows have the same ID, when we execute this query, Hibernate caches result on the first row and reuses it for the rest of the rows. 
+
+First level caching can not be turned off, there is nothing "nice" to solve this issue and have Hibernate properly working with temporal tables.
```

Comment:
```suggestion
First-level caching can not be turned off, there is "nice and easy" way to solve this issue and have Hibernate properly working with temporal tables.
```

Reply:
no, I meant that there is no nice and easy way :)

Reply:
but I understood the suggested change, so I added it